engine/lib/phases/phase_rewrite_control_flow.ml

```diff
@@ -11,31 +14,127 @@ module Make (F : Features.T) =
 
       open Ast.Make (F)
       module U = Ast_utils.Make (F)
+      module M = Ast_builder.Make (F)
       module Visitors = Ast_visitors.Make (F)
 
       module Error = Phase_utils.MakeError (struct
         let ctx = Diagnostics.Context.Phase phase_id
       end)
 
       let has_return =
+        object (_self)
+          inherit [_] Visitors.reduce as super
+          method zero = false
+          method plus = ( || )
+
+          method! visit_expr' in_loop e =
+            match e with
+            | Return _ -> true
+            | (Break _ | Continue _) when in_loop -> true
+            | _ -> super#visit_expr' in_loop e
+        end
+
+      let loop_return_type =
+        object (_self)
+          inherit [_] Visitors.reduce as super
+          method zero = (U.unit_typ, None)
+          method plus l r = if [%eq: ty] (fst l) U.unit_typ then r else l
+
+          method! visit_expr' () e =
+            match e with
+            | Return { e; witness; _ } -> (e.typ, Some witness)
+            | _ -> super#visit_expr' () e
+        end
+
+      let loop_has_cf =
         object (_self)
           inherit [_] Visitors.reduce as super
           method zero = false
           method plus = ( || )
 
           method! visit_expr' () e =
-            match e with Return _ -> true | _ -> super#visit_expr' () e
+            match e with
+            | Return _ | Break _ | Continue _ -> true
+            | Loop _ -> false
+            | _ -> super#visit_expr' () e
         end
```

Comment:
What is the difference between this and `has_return` with the `in_loop` set to `true`